Mathlib/NumberTheory/Cyclotomic/Three.lean

```diff
@@ -0,0 +1,115 @@
+https://kconrad.math.uconn.edu/blurbs/gradnumthy/kummer.pdf).
+-/
+
+open NumberField Units InfinitePlace nonZeroDivisors Polynomial
+
+namespace IsCyclotomicExtension.Rat.Three
+
+variable {K : Type*} [Field K] [NumberField K] [IsCyclotomicExtension {3} ℚ K]
+variable {ζ : K} (hζ : IsPrimitiveRoot ζ ↑(3 : ℕ+)) (u : (𝓞 K)ˣ)
+local notation3 "η" => hζ.toInteger
+local notation3 "λ" => hζ.toInteger - 1
+
+/-- Let `u` be a unit in `(𝓞 K)ˣ`, then `u ∈ {1, -1, η, -η, η^2, -η^2}`. -/
+theorem Units.mem : ↑u ∈ ({1, -1, η, -η, η ^ 2, -η ^ 2} : Set (𝓞 K)) := by
+  have hrank : rank K = 0 := by
+    dsimp only [rank]
+    rw [card_eq_nrRealPlaces_add_nrComplexPlaces, nrRealPlaces_eq_zero (n := 3) K (by decide),
+      zero_add, nrComplexPlaces_eq_totient_div_two (n := 3)]
+    rfl
+  obtain ⟨⟨x, e⟩, hxu, -⟩ := exist_unique_eq_mul_prod _ u
+  replace hxu : u = x := by
+    rw [← mul_one x.1, hxu]
+    apply congr_arg
+    rw [← Finset.prod_empty]
+    congr
+    rw [Finset.univ_eq_empty_iff, hrank]
+    infer_instance
+  obtain ⟨n, hnpos, hn⟩ := isOfFinOrder_iff_pow_eq_one.1 <| (CommGroup.mem_torsion _ _).1 x.2
+  replace hn : (↑u : K) ^ ((⟨n, hnpos⟩ : ℕ+) : ℕ) = 1 := by
+    rw [← map_pow]
+    convert map_one (algebraMap (𝓞 K) K)
+    rw_mod_cast [hxu, hn]
+    simp
+  obtain ⟨r, hr3, hru⟩ := hζ.exists_pow_or_neg_mul_pow_of_isOfFinOrder (by decide)
+    (isOfFinOrder_iff_pow_eq_one.2 ⟨n, hnpos, hn⟩)
+  replace hr : r ∈ Finset.Ico 0 3 := Finset.mem_Ico.2 ⟨by simp, hr3⟩
+  replace hru : ↑u = η ^ r ∨ ↑u = -η ^ r := by
+    rcases hru with (h | h)
+    · left; ext; exact h
+    · right; ext; exact h
+  fin_cases hr
+  all_goals{
+    rcases hru with (h | h)
+    · simp only [h, pow_zero, Set.mem_insert_iff, one_ne_zero, Set.mem_singleton_iff, false_or,
+        true_or, zero_add, pow_one, eq_neg_self_iff, true_or, or_true]
+    · simp [h, pow_zero, Set.mem_insert_iff, neg_eq_self_iff, one_ne_zero, Set.mem_singleton_iff,
+        true_or, or_true, zero_add, pow_one, neg_inj, neg_eq_self_iff]}
+
+/-- We have that `λ ^ 2 = -3 * η`. -/
+lemma lambda_sq : λ ^ 2 = -3 * η :=
+  calc λ ^ 2 = η ^ 2 + η + 1 - 3 * η := by ring
+  _ = 0 - 3 * η := by ext; simpa using hζ.isRoot_cyclotomic (by decide)
+  _ = -3 * η := by ring
+
+/-- If a unit `u` is congruent to an integer modulo `λ ^ 2`, then `u = 1` or `u = -1`.
+
+This is a special case of the so-called *Kummer's lemma*. -/
+theorem eq_one_or_neg_one_of_unit_of_congruent (hcong : ∃ n : ℤ, λ ^ 2 ∣ (u - n : 𝓞 K)) :
+    u = 1 ∨ u = -1 := by
+  replace hcong : ∃ n : ℤ, (3 : 𝓞 K) ∣ (↑u - n : 𝓞 K) := by
+    obtain ⟨n, x, hx⟩ := hcong
+    exact ⟨n, -η * x, by rw [← mul_assoc, mul_neg, ← neg_mul, ← lambda_sq, hx]⟩
+  have hζ := IsCyclotomicExtension.zeta_spec 3 ℚ K
+  have := Units.mem hζ u
+  have h2 : (hζ.pow_of_coprime 2 (by decide)).toInteger = hζ.toInteger ^ 2 := by ext; simp
+  simp only [Set.mem_insert_iff, val_eq_one, Set.mem_singleton_iff] at this
+  rcases this with (rfl | h | h | h | h | h)
```

Comment:
`fin_cases this` should work here (untested).

Reply:
It doesn't since the theorem above is wrt to `Set` and not `Finset`. I don't remember why I wrote it like that but I am modifying it using a `Finset`. Anyway `fin_cases` still doesn't work.

Reply:
It should work with a `List` if the statement is of the form `x ∈[a,b,c,...]` with an explicit list), but also with an explicit `Finset`:
```lean
import Mathlib

example {n : ℕ} (h : n ∈ ({1,2,3} : Finset ℕ)) : n ≤ 3 := by
  fin_cases h <;> omega
```

Reply:
I agree it should, but here is doesn't. The error is
```lean
tactic 'cases' failed, nested error:
dependent elimination failed, failed to solve equation
  Decidable.rec
      (fun h ↦
        (fun x ↦
            1 ::
              List.insert (-1)
                (List.insert hζ.toInteger
                  (List.insert (-hζ.toInteger) (List.insert (-hζ.toInteger - 1) [hζ.toInteger + 1]))))
          h)
      (fun h ↦
        (fun x ↦
            List.insert (-1)
              (List.insert hζ.toInteger
                (List.insert (-hζ.toInteger) (List.insert (-hζ.toInteger - 1) [hζ.toInteger + 1]))))
          h)
      (instDecidableEqBool
        (List.elem 1
          (List.insert (-1)
            (List.insert hζ.toInteger
              (List.insert (-hζ.toInteger) (List.insert (-hζ.toInteger - 1) [hζ.toInteger + 1])))))
        true) =
    ↑u :: as✝
at case List.Mem.head
```
Let me see if using a list helps (the problem is likely the fact that being able to write a finset using the `{...}` notation require some decidability instance.